File: Codes/Bingran Li/HashTable/94. Binary Tree Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        
       // solution 2: iterative 
        
        List<Integer> myList = new ArrayList();
        
        if(root==null)
            return myList;
        
        Stack<TreeNode> myStack = new Stack<TreeNode>();
        TreeNode cur = root;
        
        while(cur!=null||!myStack.empty()){
            while(cur!=null){
                myStack.push(cur);
                cur = cur.left;
            }
            cur = myStack.pop();
            myList.add(cur.val);
            cur = cur.right;
        }
        
        return myList;
        
        // solution 1: recursive
        /*
        List<Integer> myList = new ArrayList<Integer>();
        if(root==null)
            return myList;
            
        List<Integer> left = inorderTraversal(root.left);
        List<Integer> right = inorderTraversal(root.right);
        
        myList.addAll(left);
        myList.add(root.val);
        myList.addAll(right);
        
        return myList;
        */
        
    }
    
}

File: Codes/Kan Zheng/C++/Hashtable/94. Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> vec;
        if(root==NULL)
            return vec;
        stack<TreeNode*> s;
        TreeNode* cur_node=root;
        while(1)
        {
            if(cur_node->left==NULL)
            {    
                vec.push_back(cur_node->val);
                while(cur_node->right==NULL)
                {
                    if(s.empty())
                        return vec;
                    cur_node = s.top();
                    vec.push_back(cur_node->val);
                    s.pop();
                }
                cur_node = cur_node->right;
            }
            else
            {
                s.push(cur_node);
                cur_node = cur_node->left;
            }
        }
        
    }
};